include/swift/RemoteInspection/ReflectionContext.h

```diff
@@ -1202,6 +1202,33 @@ class ReflectionContext
     }
   }
 
+  /// Given a typeref, attempt to discover its class instance size. For a
+  /// version of this function that performs the same job but starting out with
+  /// an instance pointer check
+  /// MetadataReader::readInstanceStartFromClassMetadata.
+  llvm::Optional<unsigned> readInstanceStartFromTypeRef(const TypeRef *TR) {
+    size_t isaAndRetainCountSize = sizeof(StoredSize) + sizeof(long long);
+
+    const TypeRef *superclass = getBuilder().lookupSuperclass(TR);
+    if (!superclass)
+      // If there is no superclass the stat of the instance's field is right
+      // after the isa and retain fields.
+      return isaAndRetainCountSize;
+
+    auto superclassStart = readInstanceStartFromTypeRef(superclass);
+    if (!superclassStart)
+      return llvm::None;
+
+    auto *superTI = getBuilder().getTypeConverter().getClassInstanceTypeInfo(
+        superclass, *superclassStart, nullptr);
+    if (!superTI)
+      return llvm::None;
+
+    // The start of the subclass's fields is right after the super class's ones.
+    size_t start = superTI->getSize();
```

Comment:
This might need to account for padding. For example:

```
class Sup {
  var n: Int8
}

class Sub {
  var s: String
}
```

This might give you 17 (on 64-bit) but Sub's field starts at 24. Or for a more painful case:

```
class Sup {
  var n: Int
}

class Sub {
  var m: SIMD4<Double>
}
```

This will give you 24 but m is 16-byte aligned and starts at 32.

Reply:
Hmm, I tested out your examples and you're right Mike, but it looks like the where I'm using this expects the size of the superclass (except if there's no superclass, then it expects the start of the fields offset), and adjusts the alignment correctly, so I'll update the function name and comment to reflect this better.

Reply:
Sounds good!

Reply:
I ended up with `readSizeOfFieldsNotBelongingToInstanceFromTypeRef` although I'm not in love with the name. It's kind of hard to describe "the size of the superclass if there is one or the start of this instance's fields if there isn't" in a concise function name...

Reply:
The name is tough here. How about `readUnalignedFieldStartOffsetFromTypeRef`? The "unaligned" tells the caller that it's not necessarily correct, depending on the alignment of the first field.